Add EXPLAIN for simple distributed queries

marcocitus opened the pull request:
This change adds EXPLAIN for simple distributed queries, meaning queries that do not contain re-partition jobs. The EXPLAIN output consists of two parts: The distributed query and the master query. 

The distributed query part shows the remote EXPLAIN plan for either one arbitrary task, or all tasks if citus.explain_all_tasks is on. A few limitations to note:
- A remote EXPLAIN may error out when explaining a broadcast join while the shards for the small table have not yet been fetched. An error message is displayed advising to run the query first.
- When citus.explain_all_tasks is on, EXPLAIN plans are retrieved sequentially, which may take a long time for EXPLAIN ANALYZE. 
- The remote EXPLAIN is not shown by auto_explain, because auto_explain doesn't call the explain hook. We'd need to implement our own auto_explain logic to support it.

The master query part uses the regular EXPLAIN logic to show the plan of the master query by calling ExplainOnePlan. 
- During EXPLAIN ANALYZE, some tasks will be executed twice, once while fetching the remote EXPLAIN ANALYZE step and again during distributed execution.
- The execution time in EXPLAIN ANALYZE shows the execution time of the master query, which seemed more useful to understand performance than the overall execution time.

There are a few quirks as a result of limitations in the Explain functions. For example, in the JSON format, the Plan for a Task is represented as an array in an array. The outer array is unnecessary, but there is no nice way to avoid it.

A number of existing regression tests used EXPLAIN and relied on the error message saying that EXPLAIN was not supported. I first tried adding the EXPLAIN output to the expected output. However, some tests use queries that are not supported by the current implementation EXPLAIN and I didn't want to move the problem forward. I then tried changing to SELECT or CREATE TEMPORARY TABLE AS, but some tests use EXPLAIN solely to test the planner logic, while their execution actually fails. In the end, I added a citus.explain_distributed_queries option that is on by default, but can be turned off for regression tests and produces an error message as before. I did change the message for correctness, so there are a lot of changes in the regression tests. I can't think of a use-case where you might want to disable EXPLAIN, other than regression tests.

Closes #232 
Closes #15

marcocitus commented:
Example output single task:

    # explain select a.product_title, b.product_title from customer_reviews a, customer_reviews b where a.review_date = b.review_date and a.customer_id = 'AE22YDHSBFYIP' limit 10;
                                                        QUERY PLAN
    -------------------------------------------------------------------------------------------------------------------
     Distributed Query into pg_merge_job_1003
       Task Count: 14       
       Tasks Shown: Sample       
       ->  Task 
             Node: host=localhost port=9700 db_name=postgres
             ->  Limit  (cost=0.42..16.04 rows=10 width=50)
                   ->  Nested Loop  (cost=0.42..26714.71 rows=17108 width=50)
                         ->  Seq Scan on customer_reviews_102008 a  (cost=0.00..25187.24 rows=15 width=29)
                               Filter: (customer_id = 'AE22YDHSBFYIP'::text)
                         ->  Index Scan using foo on customer_reviews_102008 b  (cost=0.42..95.85 rows=598 width=29)
                               Index Cond: (review_date = a.review_date)
    
     Master Query
       ->  Limit  (cost=0.00..0.00 rows=0 width=0)
             ->  Seq Scan on pg_merge_job_1003  (cost=0.00..0.00 rows=0 width=0)
     (15 rows)

Example when workers are down:

    # explain select a.product_title, b.product_title from customer_reviews a, customer_reviews b where a.review_date = b.review_date and a.customer_id = 'AE22YDHSBFYIP' limit 10;
    WARNING:  could not connect to node localhost:9700
    DETAIL:  Client error: could not connect to server: Connection refused
        Is the server running on host "localhost" (127.0.0.1) and accepting
        TCP/IP connections on port 9700?
    
    WARNING:  could not connect to node localhost:9701
    DETAIL:  Client error: could not connect to server: Connection refused
            Is the server running on host "localhost" (127.0.0.1) and accepting
            TCP/IP connections on port 9701?
    
                                     QUERY PLAN
    -----------------------------------------------------------------------------
     Distributed Query into pg_merge_job_1004
       Task Count: 14
       Tasks Shown: Sample
       ->  Task
             Error: Could not connect to any placements

     Master Query
           ->  Limit  (cost=0.00..0.00 rows=0 width=0)
                 ->  Seq Scan on pg_merge_job_1004  (cost=0.00..0.00 rows=0 width=0)
        (9 rows)

Example when shard is not available:

    # explain select a.product_title, b.product_title from customer_reviews a, customer_reviews b where a.review_date = b.review_date and a.customer_id = 'AE22YDHSBFYIP' limit 10;
    WARNING:  could not receive query results
    DETAIL:  Client result status: PGRES_FATAL_ERROR
    
    WARNING:  could not receive query results
    DETAIL:  Client result status: PGRES_FATAL_ERROR
                                     QUERY PLAN
    -----------------------------------------------------------------------------
     Distributed Query into pg_merge_job_1005
       Task Count: 14
       Tasks Shown: Sample
       ->  Task 
             Error: Could not get plan. You may need to run the query first.

     Master Query
       ->  Limit  (cost=0.00..0.00 rows=0 width=0)
             ->  Seq Scan on pg_merge_job_1005  (cost=0.00..0.00 rows=0 width=0)
    (9 rows)

anarazel commented on src/test/regress/expected/multi_explain_0.out:
```diff
@@ -0,0 +1,334 @@
+--
```
I think we should disable psql's alignment in this file. Otherwise tiny changes in explain output will result in large output changes. Something like
```
-- temporarily disable fancy output, so output changes create less diff noise
\a\t
$test
-- restore normal output.
\a\t
```
Although I guess we can just skip enabling it again, and use it for the whole multi_explain.sql

marcocitus replied:
Good idea. I kind of wish we had that for all tests now.

anarazel commented on src/test/regress/expected/multi_hash_pruning.out:
```diff
@@ -59,120 +59,120 @@ SET client_min_messages TO DEBUG2;
- explain statements for distributed queries are currently unsupported
+                         QUERY PLAN                         
+------------------------------------------------------------
+ explain statements for distributed queries are not enabled
```
Is there any reason not to enable this by default? Seems like a pretty pointless complication for users.

marcocitus replied:
No, that's why it's enabled by default (see PR description).

anarazel commented on src/backend/distributed/planner/multi_explain.c:
```diff
@@ -99,10 +182,682 @@ MultiExplainOneQuery(Query *query, IntoClause *into, ExplainState *es,
+		{
+			/* nothing to do */
+		}
+		break;
```
It's rather weird to add braces, but have the break; outside that scope.

anarazel commented:
Hm. I think it's not a good idea to have citus.explain_distributed_queries globally off in the regression tests. We should generally write tests in an environment that's close to the setup used in production. So I'd say that we rather disable it locally in the places where we don't want the output.

lfittl commented:
I've recently had a case where this would have been quite useful - where are we currently on this PR? @anarazel @marcocitus

anarazel commented on src/backend/distributed/planner/multi_explain.c:
```diff
@@ -99,10 +182,682 @@ MultiExplainOneQuery(Query *query, IntoClause *into, ExplainState *es,
+ExplainCloseGroup(const char *objtype, const char *labelname,
+				  bool labeled, ExplainState *es)
+{
+	switch (es->format)
```
These probably should have a note that saying they're copied from postgres, and then grow a indent-off tag.

marcocitus replied:
Re-copied the functions with original alignment and added indent-off. There's a message above saying that all functions below are from explain.c

anarazel commented on src/test/regress/pg_regress_multi.pl:
```diff
@@ -94,6 +94,7 @@ ()
 push(@pgOptions, '-c', "citus.shard_max_size=300kB");
 push(@pgOptions, '-c', "citus.max_running_tasks_per_node=4");
 push(@pgOptions, '-c', "citus.expire_cached_shards=on");
+push(@pgOptions, '-c', "citus.explain_distributed_queries=off");
```
I still think we should move this away into the actual tests.

anarazel commented on src/test/regress/sql/multi_explain.sql:
```diff
@@ -0,0 +1,57 @@
+EXPLAIN (COSTS FALSE)
+	DELETE FROM lineitem
+	WHERE l_orderkey = 1 AND l_partkey = 0;
+
```
Should have tests for the different executors, particularly for router executor, modification, EXECUTE, CREATE TABLE AS, EXECUTE. I think it's ok if we don't support all of them, but we should make sure we fail gracefully.

marcocitus replied:
Added extra tests. I need to think a bit about EXECUTE.

marcocitus replied:
It seems that EXPLAIN EXECUTE doesn't go through the explain hook. It will just show the local plan right now.

anarazel replied:
Gah. Who designed this hook. Horrible.

anarazel commented:
@marcocitus There's actually pending comments to be addressed and a rebase.

anarazel commented on src/test/regress/expected/multi_explain_0.out:
```diff
@@ -0,0 +1,334 @@
+	SELECT avg(l_linenumber) FROM lineitem WHERE l_orderkey > 9030;
+                        QUERY PLAN                         
+-----------------------------------------------------------
+ Distributed Query into pg_merge_job_0064
```
I'm not sure we can realistically address this right now - and perhaps it doesn't make sense as we want to get rid of the temp tables anyway - but I think displaying the temp table name here is a bad idea. Will just introduce pointless variance, and it's not really giving the user more information.

marcocitus replied:
The temporary table will show up in the master query plan anyway, so showing it twice doesn't introduce any new variance. I think it clarifies the link between the distributed query and the master query a bit.

marcocitus commented:
@anarazel 
I reworked the multi_hash_pruning tests, because the queries were no longer plannable after the router planner was added, because it assumes hash-partitioned tables are evenly distributed. 
multi_hash_pruning predates master_create_worker_shards and created mock shard metadata manually with gaps and overlaps. However, the tests still passed because the old EXPLAIN logic only used the physical planner not the router planner. I changed the tests to use master_create_worker_shards. Since the queries were fairly trivial, it also seemed better to stop using EXPLAIN altogether.

mtuncer commented:
@marcocitus if you need you can by pass router executor by adding a dummy where clause that contains the partition column

```select .... from .... where partitionColumn = xxx``` could be converted to 
```select .... from .... where (partitionColumn = xxx and partitionColumn != yyy)```. Query is still the same but it becomes not router plannable

anarazel commented:
I'm wondering if we shouldn't show the slowest task instead of a random task. In practice that's likely going to be a *lot* more useful. But I grant it's not necessary for the first iteration.

anarazel commented on src/backend/distributed/planner/multi_explain.c:
```diff
@@ -99,10 +173,655 @@ MultiExplainOneQuery(Query *query, IntoClause *into, ExplainState *es,
+		es->indent -= 1;
+	}
+
+	routerExecutablePlan = RouterExecutablePlan(multiPlan, TaskExecutorType);
```
Why aren't we supporting router executor properly? Especially with dynamic selection that seems rather confusing for users?

marcocitus replied:
What does properly mean? A router executor query has no master query, so it's not displayed.

anarazel commented on src/backend/distributed/planner/multi_explain.c:
```diff
@@ -99,10 +173,655 @@ MultiExplainOneQuery(Query *query, IntoClause *into, ExplainState *es,
+
+/*
+ * ExplainJob shows the EXPLAIN output for a Job in the physical plan of
+ * a distributed query by showing the first remote EXPLAIN for the first
```
Description is out of date wrt. ExplainAllTasks

marcocitus replied:
Fixed

anarazel commented on src/backend/distributed/planner/multi_explain.c:
```diff
@@ -99,10 +173,655 @@ MultiExplainOneQuery(Query *query, IntoClause *into, ExplainState *es,
+		es->indent += 1;
+	}
+
+	ExplainJob(workerJob, es);
```
I suspect that we should, here or closeby, add a text property explaining which executor was used. That's not clear anymore with dynamic selection, and it's likely not going to get any clearer.

marcocitus replied:
I agree. I'll add that.

anarazel commented on src/backend/distributed/planner/multi_explain.c:
```diff
@@ -99,10 +173,655 @@ MultiExplainOneQuery(Query *query, IntoClause *into, ExplainState *es,
+	if (list_length(workerJob->dependedJobList) > 0)
+	{
+		appendStringInfo(es->str, "explain statements for complex queries ");
 		appendStringInfo(es->str, "are currently unsupported\n");
```
We chose to accept this restriction because of the amount of effort, and the comparatively low demand for complex task tracker stuff?

anarazel replied:
Couldn't we at least show a) executor, b) master query c) number of tasks or something like that?

marcocitus replied:
Yes. It's a bit tricky because the queries use intermediate tables that are only created during re-partitioning, but we don't want to actually run the re-partitioning job when doing a regular EXPLAIN, so we'd need some infrastructure for creating the tables. There are no fundamental limitations preventing us from doing that. We originally marked that as "phase 2" of EXPLAIN, with the most critical parts (this PR) being phase 1. 

I suppose we can show something, but it doesn't seem like it would convey very useful information.

anarazel replied:
Executor  would be useful, the rest seems sensible.

marcocitus replied:
Ok, added some basic output showing the number of map and merge tasks in each MapMergeJob and the task count in the top-level job.

anarazel commented on src/test/regress/output/multi_master_delete_protocol.source:
```diff
@@ -70,7 +70,7 @@ SELECT count(*) FROM customer_delete_protocol;
 SELECT master_create_empty_shard('customer_delete_protocol');
  master_create_empty_shard 
 ---------------------------
```
Seems to suggest we should rather have this test reset the sequence explicitly.

marcocitus replied:
Changed it to just hide the shard id.

anarazel commented on src/test/regress/expected/multi_explain.out:
```diff
@@ -0,0 +1,356 @@
+        ->  HashAggregate
+              Group Key: l_quantity
+              ->  Seq Scan on lineitem_102010 lineitem
+
```
Why is there a newline here?

marcocitus replied:
Removed

anarazel commented on src/backend/distributed/planner/multi_explain.c:
```diff
@@ -99,10 +173,655 @@ MultiExplainOneQuery(Query *query, IntoClause *into, ExplainState *es,
+	}
+	else
+	{
+		ExplainPropertyText("Tasks Shown", "Sample", es);
```
`Sample` isn't really accurate - isn't this most of the the time going to be deterministic task?

marcocitus replied:
I'm not in love with this word. I previously had "Single", which just seemed weird, and "First", which didn't have any meaning to the user. "Sample" is at least accurate (a sample is not necessarily random), but open to alternatives.

marcocitus replied:
Maybe I can just remove the whole "Tasks shown" thing, since it doesn't add that much.

anarazel replied:
Just say: 'One of X' vs. 'All'?

marcocitus replied:
Works for me.

anarazel commented:
So, the biggest problem I have with this approach is that it has to re-implement a significant portion of task execution (c.f. ExplainTaskPlacement). The desire to not invest too much effort in that duplication leads to problems like ExplainAllTasks being slow due to sequential execution, broadcast joins not being supported, et al.

Especially the non-parallelized execution means you'll often get very different timings from EXPLAIN ANALYZE, than when running the queries for real.

I do wonder if we shouldn't instead rely almost entirely on the normal execution, and determine in the relevant places that we're currently doing EXPLAIN or ANALYZE, and stash the relevant data in a hashtable or such. Then, *after* that, iterate through the tree and grab the details from the hashtable and print the results. 

That approach would be closer, imnsho, to the way postgres itself implements EXPLAIN (be it with ANALYZE or without). It'll require some minor additions of intelligence in the executors when issuing queries, about using explain or not.


I'm fairly convinced that that approach would be better. But I'm also fairly convinced that it's unrealistic to get that into 5.1. So I think it's reasonable to go with this approach, and consider refactoring this. Although that has sunken-cost fallacy type risk of never going tho other way.

marcocitus commented:
I can see where you're coming from since integrating explain into the executors is the first design I considered as well, but I found that it requires major changes to each of the executors. Doing it just for the real-time executor would involve adding several new states with different transitions depending on whether it is a regular select, explain, or explain analyze, and various other changes. You'd also need some kind of backdoor API to pass parameters and results to/from the executors. I'm not even sure how it would work for task-tracker executor, and there's also the router executor. It leads to poor separation of concerns and would be hard to maintain. I abandoned this approach in favour of one that keeps the explain logic in one place, which I think is the right approach now and in the future.

Some level of duplication is inherent to modularity. However, poor separation of concerns is much costlier to maintain than duplication.

In the future, I do think it's worth adding some instrumentation to executors to watch for task execution times such that we can find the slowest task. We could also add end-to-end execution times to the EXPLAIN output. I wanted to keep the initial code simple.

> Especially the non-parallelized execution means you'll often get very different timings from EXPLAIN ANALYZE, than when running the queries for real.

There is some value in running tasks in isolation. Execution times will almost certainly not be accurate or representative if you run EXPLAIN ANALYZE tasks in parallel since it doesn't just add overhead but the overhead also causes additional CPU contention, especially if the executor mixes EXPLAIN ANALYZE with regular SELECTs. The numbers end up being fairly useless. Some tasks have high execution times for no other reason than being co-located with other tasks that require a lot of CPU time. It may be a lot useful to be able to see which tasks require a lot of CPU time, for which they'd need to be run separately. Clearly, you would still get the same benefit if you ran one task per worker in parallel, but that's also more complex and we don't have existing code for it.

I think one thing I could do is separate fetching and processing the remote explains a bit more so it would be easier to parallelize the latter in the future. What do you think?

marcocitus commented:
I've simplified the way EXPLAIN plans are fetched a bit by calling ExecuteRemoteQuery, and separated it from processing the EXPLAIN plans, which will make it easier to pick a different strategy for fetching the plans in the future.

anarazel commented:
> I can see where you're coming from since integrating explain into the
> executors is the first design I considered as well, but I found that
> it requires major changes to each of the executors.

Yea, it'd require some additional work, no denying that.

> Doing it just for the real-time executor would involve adding several
> new states with different transitions depending on whether it is a
> regular select, explain, or explain analyze, and various other
> changes. You'd also need some kind of backdoor API to pass parameters
> and results to/from the executors. I'm not even sure how it would work
> for task-tracker executor, and there's also the router executor. It
> leads to poor separation of concerns and would be hard to maintain. I
> abandoned this approach in favour of one that keeps the explain logic
> in one place, which I think is the right approach now and in the
> future.

I can't agree here. This approach will really only work to simulate a
subset of actual query execution, and the better our query execution
gets, the smaller this will be.  I also don't think it'd be quite that
bad to implement.

But it doesn't really matter for the moment. This is a lot better than
the current state in master.


> In the future, I do think it's worth adding some instrumentation to
> executors to watch for task execution times such that we can find the
> slowest task. We could also add end-to-end execution times to the
> EXPLAIN output. I wanted to keep the initial code simple.

Yes, that makes sense.


> > Especially the non-parallelized execution means you'll often get
> > very different timings from EXPLAIN ANALYZE, than when running the
> > queries for real.

> There is some value in running tasks in isolation. Execution times
> will almost certainly not be accurate or representative if you run
> EXPLAIN ANALYZE tasks in parallel since it doesn't just add overhead
> but the overhead also causes additional CPU contention, especially if
> the executor mixes EXPLAIN ANALYZE with regular SELECTs. The numbers
> end up being fairly useless.

I don't think that's really true. If you execute some IO intensive query
you're likely see all tasks in an explain being very fast, but if you
execute the whole thing at once - as it normally happens - you'll be
terribly IO bound because of concurrent random IO.   Similar for CPU,
the fact that you get various forms of contention *is* relevant.


> It may be a lot useful to be able to see which tasks require a lot of CPU time, for
> which they'd need to be run separately.

Nearly all kernels track CPU time on a per process basis
(c.f. getrusage() and log_statement_stats), we should be able to
integrate whether the tasks run in isolation or not.

Clearly, you would still get the
same benefit if you ran one task per worker in parallel, but that's also
more complex and we don't have existing code for it.


Andres

anarazel commented on src/backend/distributed/planner/multi_explain.c:
```diff
@@ -99,10 +175,728 @@ MultiExplainOneQuery(Query *query, IntoClause *into, ExplainState *es,
+	}
+	else
+	{
+		ExplainPropertyText("Executor", "Other", es);
```
I'd use a switch, because that nicely warns if a new executor type has been added. But whatever.

anarazel commented:
Besides that minor comment, this looks reasonable: :shipit:
